shared/message.h

```diff
@@ -22,7 +23,34 @@ enum class cmdtype : std::size_t {
     LAST
 };
 
-struct message_register {
+template<typename Impl>
+struct base_message {
+    Impl &self() { return static_cast<Impl &>(*this); }
+    bool base_validate(const nibashared::sessionstate &session) { return self().validate(session); }
+    nlohmann::json base_create_response() {
+        auto j = self().create_response();
+        j["success"] = success;
+        return j;
+    }
+    nlohmann::json base_create_request() {
+        auto j = self().create_request();
+        j["type"] = self().type;
+        return j;
+    }
+    void base_merge_response(const nlohmann::json &j) {
+        j.at("success").get_to(success);
+        self().merge_response(j);
+    }
+    void base_from_request(const nlohmann::json &j) { self().from_request(j); }
+
+    bool success = false;
+};
+
+template<typename T>
+using IsMessage = typename std::enable_if<std::is_base_of<base_message<T>, T>::value>::type;
```

Comment:
Can you explain this statement?

Reply:
this checks if `T` is derived from `base_message<T>` which can be used as 
`typename = IsMessage<Message>` in a template